File: mcp-list/packages/EventAnalyzer/src/utils/type_checker.py
"""类型检查器
用于推断和校验埋点字段的类型
"""

from typing import Any, Literal


TypeName = Literal["NUMBER", "STRING", "BOOL", "LIST", "OBJECT", "NULL", "UNKNOWN"]


class TypeChecker:
    """类型检查和推断工具"""

    @staticmethod
    def infer_type(value: Any) -> TypeName:
        """
        推断值的类型

        Args:
            value: 待推断的值

        Returns:
            类型名称（NUMBER/STRING/BOOL/LIST/OBJECT/NULL/UNKNOWN）
        """
        if value is None:
            return "NULL"

        if isinstance(value, bool):
            return "BOOL"

        if isinstance(value, (int, float)):
            return "NUMBER"

        if isinstance(value, str):
            return "STRING"

        if isinstance(value, list):
            return "LIST"

        if isinstance(value, dict):
            return "OBJECT"

        return "UNKNOWN"

    @staticmethod
    def type_matches(expected: TypeName, actual: TypeName) -> bool:
        """
        检查类型是否匹配

        Args:
            expected: 期望的类型
            actual: 实际的类型

        Returns:
            是否匹配
        """
        # 完全匹配
        if expected == actual:
            return True

        # NULL 可以匹配任何类型（可选字段）
        if actual == "NULL":
            return True

        # BOOL 在某些情况下可以视为 NUMBER (0/1)
        if expected == "NUMBER" and actual == "BOOL":
            return True

        # NUMBER 在某些情况下可以转换为 STRING
        if expected == "STRING" and actual == "NUMBER":
            return True

        return False

    @staticmethod
    def validate_value(value: Any, expected_type: TypeName) -> tuple[bool, str]:
        """
        验证值是否符合期望的类型

        Args:
            value: 待验证的值
            expected_type: 期望的类型

        Returns:
            (是否有效, 错误信息)
        """
        actual_type = TypeChecker.infer_type(value)

        if TypeChecker.type_matches(expected_type, actual_type):
            return True, ""

        error_msg = f"类型不匹配: 期望 {expected_type}, 实际 {actual_type}"
        return False, error_msg

    @staticmethod
    def is_number_like(value: Any) -> bool:
        """检查值是否可以被视为数字"""
        if isinstance(value, (int, float)):
            return True

        if isinstance(value, bool):
            return True

        if isinstance(value, str):
            try:
                float(value)
                return True
            except ValueError:
                return False

        return False

    @staticmethod
    def is_bool_like(value: Any) -> bool:
        """检查值是否可以被视为布尔值"""
        if isinstance(value, bool):
            return True

        if isinstance(value, (int, float)):
            return value in (0, 1)

        if isinstance(value, str):
            return value.lower() in ("true", "false", "0", "1")

        return False

File: mcp-list/packages/EventAnalyzer-node/src/utils/typeChecker.ts
/**
 * 类型检查器
 * 用于推断值的类型并检查类型匹配
 */

export type TypeName = 'NUMBER' | 'STRING' | 'BOOL' | 'LIST' | 'OBJECT' | 'NULL' | 'UNKNOWN';

export class TypeChecker {
  /**
   * 推断值的类型
   */
  static inferType(value: any): TypeName {
    if (value === null || value === undefined) {
      return 'NULL';
    }

    if (typeof value === 'number') {
      return 'NUMBER';
    }

    if (typeof value === 'boolean') {
      return 'BOOL';
    }

    if (typeof value === 'string') {
      return 'STRING';
    }

    if (Array.isArray(value)) {
      return 'LIST';
    }

    if (typeof value === 'object') {
      return 'OBJECT';
    }

    return 'UNKNOWN';
  }

  /**
   * 检查类型是否匹配（支持隐式转换）
   * @param expected 期望的类型
   * @param actual 实际的类型
   * @param value 实际的值（用于隐式转换检查）
   */
  static typeMatches(expected: TypeName, actual: TypeName, value?: any): boolean {
    // 完全匹配
    if (expected === actual) {
      return true;
    }

    // NULL 类型特殊处理
    if (actual === 'NULL') {
      return false;
    }

    // 隐式转换规则
    if (expected === 'NUMBER' && actual === 'STRING' && value !== undefined) {
      // 字符串可以转换为数字
      const num = Number(value);
      return !isNaN(num);
    }

    if (expected === 'BOOL' && actual === 'STRING' && value !== undefined) {
      // "0"/"1" 或 "true"/"false" 可以转换为布尔值
      const lowerValue = String(value).toLowerCase();
      return ['0', '1', 'true', 'false'].includes(lowerValue);
    }

    if (expected === 'BOOL' && actual === 'NUMBER' && value !== undefined) {
      // 0/1 可以转换为布尔值
      return value === 0 || value === 1;
    }

    if (expected === 'STRING') {
      // 任何类型都可以转换为字符串
      return true;
    }

    return false;
  }

  /**
   * 获取类型的友好名称
   */
  static getTypeName(type: TypeName): string {
    const typeMap: Record<TypeName, string> = {
      NUMBER: '数字',
      STRING: '字符串',
      BOOL: '布尔值',
      LIST: '数组',
      OBJECT: '对象',
      NULL: '空值',
      UNKNOWN: '未知类型',
    };
    return typeMap[type] || type;
  }

  /**
   * 验证枚举值
   * @param value 要验证的值
   * @param enumValues 枚举值映射 { key: description }
   * @returns 是否是有效的枚举值
   */
  static validateEnum(value: any, enumValues: Record<string, string>): boolean {
    if (!enumValues || Object.keys(enumValues).length === 0) {
      return true;
    }

    const valueStr = String(value);
    return Object.keys(enumValues).includes(valueStr);
  }
}
